gr-soapy/lib/block_impl.cc

```diff
@@ -0,0 +1,682 @@
+        SoapySDR::Kwargs settings_kwargs =
+            SoapySDR::KwargsFromString(other_settings[arg_idx]);
+
+        /*
+         * Before applying the setting, make a last check if the setting
+         * is a valid gain
+         */
+        std::vector<std::string> gains = d_device->listGains(d_direction, channel);
+        for (auto i : gains) {
+            SoapySDR::Kwargs::iterator iter = settings_kwargs.find(i);
+            if (iter != settings_kwargs.end()) {
+                d_device->setGain(d_direction, channel, i, std::stod(iter->second));
+                /*
+                 * As only valid settings are applied, we remove it so it does not
+                 * recognized as an invalid setting
+                 */
+                settings_kwargs.erase(iter);
+            }
+        }
+
+        for (auto i : settings_kwargs) {
+            bool found = false;
+            for (auto arg : supported_settings) {
+                if (arg.key == i.first) {
+                    found = true;
+                    break;
+                }
+            }
+            if (found) {
+                d_device->writeSetting(d_direction, channel, i.first, i.second);
+            } else {
+                std::string msg = name() + ": Unsupported setting " + i.first +
+                                  " for channel " + std::to_string(channel);
+                throw std::invalid_argument(msg);
+            }
+        }
+    }
+
+    message_port_register_in(pmt::mp("cmd"));
+    set_msg_handler(pmt::mp("cmd"),
+                    [this](pmt::pmt_t pmt) { this->msg_handler_cmd(pmt); });
+    register_msg_cmd_handler(CMD_FREQ_KEY, [this](pmt::pmt_t val, size_t channel) {
+        this->cmd_handler_frequency(val, channel);
+    });
+    register_msg_cmd_handler(CMD_GAIN_KEY, [this](pmt::pmt_t val, size_t channel) {
+        this->cmd_handler_gain(val, channel);
+    });
+
+    register_msg_cmd_handler(CMD_RATE_KEY, [this](pmt::pmt_t val, size_t channel) {
+        this->cmd_handler_samp_rate(val, channel);
+    });
+    register_msg_cmd_handler(CMD_BW_KEY, [this](pmt::pmt_t val, size_t channel) {
+        this->cmd_handler_bw(val, channel);
+    });
+    register_msg_cmd_handler(CMD_ANTENNA_KEY, [this](pmt::pmt_t val, size_t channel) {
+        this->cmd_handler_antenna(val, channel);
+    });
+}
+
+bool block_impl::start()
+{
+    d_stream = d_device->setupStream(
+        d_direction, d_soapy_type, d_channels, SoapySDR::KwargsFromString(d_stream_args));
+    d_mtu = d_device->getStreamMTU(d_stream);
+
+    /* This limits each work invocation to MTU transfers */
+    if (d_mtu > 0) {
+        set_max_noutput_items(d_mtu);
+    } else {
+        set_max_noutput_items(1024);
+    }
+
+    d_device->activateStream(d_stream);
+    return true;
+}
+
+bool block_impl::stop()
+{
+    if (!d_stopped) {
+        d_device->closeStream(d_stream);
+        SoapySDR::Device::unmake(d_device);
```

Comment:
Rather than leave dangling pointers, it would be neater to set these two pointers to `nullptr` after they're done.